File: server/src/main/java/org/cleanas2/pipeline/Pipeline.java
package org.cleanas2.pipeline;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.cleanas2.server.ServerSession;

import java.util.ArrayList;
import java.util.List;

public class Pipeline<TContext extends PipelineContext> {
    private static final Log logger = LogFactory.getLog(Pipeline.class.getSimpleName());

    private final List<PipelineTask<TContext>> stages = new ArrayList<>(10);
    private final List<PipelineTask<TContext>> onDone = new ArrayList<>(2);
    private final List<FailureTask<TContext>> onFailure = new ArrayList<>(2);

    public void step(PipelineTask<TContext> inst) {
        stages.add(inst);
    }

    public void step(Class<? extends PipelineTask<TContext>> klass) {
        step(ServerSession.getSession().getInstance(klass));
    }

    @SafeVarargs
    public final void step(Class<? extends PipelineTask<TContext>>... klasses) {
        for (Class<? extends PipelineTask<TContext>> c : klasses) {
            step(ServerSession.getSession().getInstance(c));
        }
    }

    public void done(PipelineTask<TContext> inst) {
        onDone.add(inst);
    }

    public void done(Class<? extends PipelineTask<TContext>> klass) {
        done(ServerSession.getSession().getInstance(klass));
    }

    public void fail(FailureTask<TContext> inst) {
        onFailure.add(inst);
    }

    public void fail(Class<? extends FailureTask<TContext>> klass) {
        fail(ServerSession.getSession().getInstance(klass));
    }

    public void run(TContext context) throws Exception {

        try {
            for (PipelineTask<TContext> h : stages) {
                logStage("Running stage", h);
                h.process(context);
                if (context.isTerminated()) {
                    break;
                }
            }
        } catch (Exception e) {
            for (FailureTask<TContext> h : onFailure) {
                logStage("Running failure handler", h);
                h.process(context, e);
            }
        }

        for (PipelineTask<TContext> h : onDone) {
            try {
                logStage("Running done handler", h);
                h.process(context);
            } catch (Exception e) {
                logger.error("Failure processing 'On Complete' action", e);
            }
        }
    }

    private static void logStage(String message, Object obj) {
        Class c = obj.getClass();
        String stage = c.isAnonymousClass() ? "(anon class)" : c.getSimpleName();
        logger.debug(message + ": " + stage);
    }

}

File: server/src/main/java/org/cleanas2/pipeline/Pipeline.kt
package org.cleanas2.pipeline

import org.apache.commons.logging.Log
import org.apache.commons.logging.LogFactory
import org.cleanas2.server.ServerSession

import java.util.ArrayList

class Pipeline<TContext : PipelineContext> {

    private val stages = ArrayList<PipelineTask<TContext>>(10)
    private val onDone = ArrayList<PipelineTask<TContext>>(2)
    private val onFailure = ArrayList<FailureTask<TContext>>(2)

    fun step(inst: PipelineTask<TContext>) {
        stages.add(inst)
    }

    fun step(klass: Class<out PipelineTask<TContext>>) {
        step(ServerSession.session!!.getInstance(klass))
    }

    @SafeVarargs
    fun step(vararg klasses: Class<out PipelineTask<TContext>>) {
        for (c in klasses) {
            step(ServerSession.session!!.getInstance(c))
        }
    }

    fun done(inst: PipelineTask<TContext>) {
        onDone.add(inst)
    }

    fun done(klass: Class<out PipelineTask<TContext>>) {
        done(ServerSession.session!!.getInstance(klass))
    }

    fun fail(inst: FailureTask<TContext>) {
        onFailure.add(inst)
    }

    fun fail(klass: Class<out FailureTask<TContext>>) {
        fail(ServerSession.session!!.getInstance(klass))
    }

    @Throws(Exception::class)
    fun run(context: TContext) {

        try {
            for (h in stages) {
                logStage("Running stage", h)
                h.process(context)
                if (context.isTerminated) {
                    break
                }
            }
        } catch (e: Exception) {
            for (h in onFailure) {
                logStage("Running failure handler", h)
                h.process(context, e)
            }
        }

        for (h in onDone) {
            try {
                logStage("Running done handler", h)
                h.process(context)
            } catch (e: Exception) {
                logger.error("Failure processing 'On Complete' action", e)
            }

        }
    }

    companion object {
        private val logger = LogFactory.getLog(Pipeline::class.java.simpleName)

        private fun logStage(message: String, obj: Any) {
            val c = obj.javaClass
            val stage = if (c.isAnonymousClass) "(anon class)" else c.simpleName
            logger.debug(message + ": " + stage)
        }
    }

}
